web/src/main/java/org/wso2/testgrid/web/api/SSOService.java

```
@@ -0,0 +1,153 @@
+            String msg = "Error occurred while unmarshalling SAMLResponse.";
+            logger.error(msg, e);
+            return Response.serverError().entity(
+                    new ErrorResponse.ErrorResponseBuilder().setMessage(msg + e.getMessage()).build()).build();
+        }
+
+        try {
+            validateSignature(saml2Response.getSignature());
+            request.getSession();
+            return Response.status(Response.Status.FOUND).
+                    header(HttpHeaders.LOCATION, Constants.WEBAPP_CONTEXT).
+                    type(MediaType.TEXT_PLAIN).build();
+        } catch (SSOAgentException e) {
+            String msg = "Observed an attempt with invalid SAMLResponse.";
+            logger.error(msg, e);
+            return Response.status(Response.Status.UNAUTHORIZED).entity(msg + e.getMessage()).build();
+        }
+    }
+
+    /**
+     * Validates the XML Signature object
+     *
+     * @param signature XMLObject
+     * @throws TestGridException if either SignatureProfile or Signature validation is failed.
+     */
+    private void validateSignature(XMLObject signature) throws SSOAgentException, TestGridException {
+
+        SignatureImpl signImpl = (SignatureImpl) signature;
+        try {
+            SAMLSignatureProfileValidator signatureProfileValidator = new SAMLSignatureProfileValidator();
+            signatureProfileValidator.validate(signImpl);
+        } catch (ValidationException e) {
+            throw new TestGridException("Signature do not confirm to SAML signature profile. Possible XML Signature " +
+                    "Wrapping Attack!", e);
+        }
+
+        try {
+            SignatureValidator validator = new SignatureValidator(
+                    new X509CredentialImpl(getX509Credential()));
+            validator.validate(signImpl);
+        } catch (ValidationException e) {
+            throw new TestGridException("Signature validation failed for SAML2 Element.");
```

*Signature validation failed for the incoming SAML2 response. You have swallowed the actual exception here. Can we at least send it to debug logs?